X509.xs

```diff
@@ -893,23 +893,30 @@ pubkey(x509)
 
   pkey_id = EVP_PKEY_base_id(pkey);
   if (pkey_id == EVP_PKEY_RSA) {
-    RSA *rsa_pkey;
+    const RSA *rsa_pkey;
 
     rsa_pkey = EVP_PKEY_get0_RSA(pkey);
     PEM_write_bio_RSAPublicKey(bio, rsa_pkey);
 
   } else if (pkey_id == EVP_PKEY_DSA) {
-    DSA *dsa_pkey;
+    const DSA *dsa_pkey;
 
     dsa_pkey = EVP_PKEY_get0_DSA(pkey);
-
-    PEM_write_bio_DSA_PUBKEY(bio, dsa_pkey);
+    /* In openssl v3, EVP_PKEY_get0_DSA came to return "const DSA*" type.
+     * However PEM_write_bio_DSA_PUBKEY still needs non-const DSA*.
+     * So we should force to cast dsa_pkey to "DSA*" here.
+     */
+    PEM_write_bio_DSA_PUBKEY(bio, (DSA*)dsa_pkey);
 #ifndef OPENSSL_NO_EC
   } else if (pkey_id == EVP_PKEY_EC ) {
-    EC_KEY *ec_pkey;
+    const EC_KEY *ec_pkey;
 
     ec_pkey = EVP_PKEY_get0_EC_KEY(pkey);
-    PEM_write_bio_EC_PUBKEY(bio, ec_pkey);
+    /* In openssl v3, EVP_PKEY_get0_EC_KEY came to return "const EC_KEY*" type.
+     * However PEM_write_bio_EC_PUBKEY still needs non-const EC_KEY*.
+     * So we should force to cast ec_pkey to "EC_KEY*" here.
+     */
+    PEM_write_bio_EC_PUBKEY(bio, (EC_KEY*)ec_pkey);
```

Comment:
This all makes sense.  I had done most of it in a branch I was looking at for v3 support as well.  Adding const does not seem to break openssl 1.1 support

Reply:
Thanks